src/ethereum/osaka/bal_utils.py

```diff
@@ -0,0 +1,338 @@
+    MAX_SLOTS,
+    MAX_ACCOUNTS,
+    MAX_CODE_SIZE,
+)
+
+
+def compute_bal_hash(bal: BlockAccessList) -> Hash32:
+    """
+    Compute the hash of a Block Access List.
+    
+    The BAL is SSZ-encoded and then hashed with keccak256.
+    
+    Parameters
+    ----------
+    bal :
+        The Block Access List to hash.
+        
+    Returns
+    -------
+    hash :
+        The keccak256 hash of the SSZ-encoded BAL.
+    """
+    bal_bytes = ssz_encode_block_access_list(bal)
+    return keccak256(bal_bytes)
+
+
+def ssz_encode_uint(value: Union[int, Uint], size: int) -> bytes:
+    """Encode an unsigned integer as SSZ (little-endian)."""
+    if isinstance(value, Uint):
+        value = int(value)
+    return value.to_bytes(size, 'little')
+
+
+def ssz_encode_bytes(data: bytes) -> bytes:
+    """Encode fixed-size bytes as SSZ."""
+    return data
+
+
+def ssz_encode_list(items: tuple, encode_item_fn, max_length: int = None) -> bytes:
+    """Encode a list/tuple as SSZ with optional max length."""
+    # For variable-length lists, we need offset encoding
+    # First, encode the list length
+    result = bytearray()
+    
+    # If max_length is specified, this is a variable-length list
+    if max_length is not None:
+        # Variable-length lists use offset encoding
+        # First 4 bytes: offset to start of data
+        item_count = len(items)
+        if item_count == 0:
+            # Empty list is encoded as just the 4-byte offset pointing to itself
+            return ssz_encode_uint(4, 4)
+        
+        # Calculate if items are fixed or variable size
+        first_item_encoded = encode_item_fn(items[0]) if items else b''
+        is_fixed_size = all(len(encode_item_fn(item)) == len(first_item_encoded) for item in items)
+        
+        if is_fixed_size:
+            # Fixed-size elements: concatenate directly
+            for item in items:
+                result.extend(encode_item_fn(item))
+        else:
+            # Variable-size elements: use offset encoding
+            # Reserve space for offsets
+            offset_start = 4 * item_count
+            data_section = bytearray()
+            
+            for item in items:
+                # Write offset
+                result.extend(ssz_encode_uint(offset_start + len(data_section), 4))
+                # Encode item data
+                item_data = encode_item_fn(item)
+                data_section.extend(item_data)
+            
+            result.extend(data_section)
+    else:
+        # Fixed-length list/tuple: just concatenate
+        for item in items:
+            result.extend(encode_item_fn(item))
+    
+    return bytes(result)
+
+
+def ssz_encode_storage_change(change: StorageChange) -> bytes:
+    """Encode a StorageChange as SSZ."""
+    result = bytearray()
+    result.extend(ssz_encode_uint(change.tx_index, 2))  # TxIndex as uint16
+    result.extend(ssz_encode_bytes(change.new_value))   # StorageValue as Bytes32
+    return bytes(result)
```

Comment:
Definitely not a fully thought out idea, but would this be more readable:

```suggestion
    return (
        ssz_encode_uint(change.tx_index, 2))  # TxIndex as uint16
        + ssz_encode_bytes(change.new_value))   # StorageValue as Bytes32
    )
```